### DISCUSSION TITLE
Vault not accessible after update to latest release. 502 Error

### DISCUSSION THREAD

**Discussion by giteke** [dani-garcia/vaultwarden#3828] [Q&A] [answered]
Hello, I updated my long running vaultwarden & Caddy setup yesterday and now I can not access the web vault. Any pointers would be appreciated.

**docker-compose.yml**
`version: '3'

services:
  vaultwarden:
    image: vaultwarden/server:latest
    container_name: vaultwarden
    restart: always
    environment:
      DOMAIN: "https://vault.fqdn.com"  # Your domain; vaultwarden needs to know it's https to work properly with attachments
    volumes:
      - ./vw-data:/data

  caddy:
    image: caddy:2
    container_name: caddy
    restart: always
    ports:
      - 80:80  # Needed for the ACME HTTP-01 challenge.
      - 443:443
    volumes:
      - ./Caddyfile:/etc/caddy/Caddyfile:ro
      - ./caddy-config:/config
      - ./caddy-data:/data
    environment:
      DOMAIN: "https://vault.fqdn.com"  # Your domain.
      EMAIL: "letsencrypt@fqdn.com"                 # The email address to use for ACME registration.
      LOG_FILE: "/data/access.log"`

**curl -v https://vault.fqdn.com**
`*   Trying 34.75.***.***:443...
* Connected to vault.fqdn.com (34.75.***.***) port 443 (#0)
* ALPN: offers h2,http/1.1
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /etc/ssl/cert.pem
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-CHACHA20-POLY1305-SHA256
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=vault.fqdn.com
*  start date: Aug 17 10:20:55 2023 GMT
*  expire date: Nov 15 10:20:54 2023 GMT
*  subjectAltName: host "vault.fqdn.com" matched cert's "vault.fqdn.com"
*  issuer: C=US; O=Let's Encrypt; CN=R3
*  SSL certificate verify ok.
* using HTTP/2
* h2h3 [:method: GET]
* h2h3 [:path: /]
* h2h3 [:scheme: https]
* h2h3 [:authority: vault.fqdn.com]
* h2h3 [user-agent: curl/7.88.1]
* h2h3 [accept: */*]
* Using Stream ID: 1 (easy handle 0x15b011400)
> GET / HTTP/2
> Host: vault.fqdn.com
> user-agent: curl/7.88.1
> accept: */*
> 
< HTTP/2 502 
< alt-svc: h3=":443"; ma=2592000
< server: Caddy
< content-length: 0
< date: Fri, 01 Sep 2023 13:38:26 GMT
< 
* Connection #0 to host vault.fqdn.com left intact`

**caddy.log**
`{"level":"error","ts":1693576459.1053498,"logger":"http.log.access.log0","msg":"handled request","request":{"remote_ip":"redacted","remote_port":"58128","client_ip":"redacted","proto":"HTTP/2.0","method":"GET","host":"vault.fqdn.com","uri":"/","headers":{"User-Agent":["Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/116.0"],"Accept":["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8"],"Accept-Language":["en-US,en;q=0.5"],"Dnt":["1"],"Sec-Fetch-Mode":["navigate"],"Sec-Fetch-Site":["none"],"Sec-Fetch-User":["?1"],"Accept-Encoding":["gzip, deflate, br"],"Upgrade-Insecure-Requests":["1"],"Sec-Fetch-Dest":["document"],"Te":["trailers"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","server_name":"vault.fqdn.com"}},"bytes_read":0,"user_id":"","duration":0.045518139,"size":0,"status":502,"resp_headers":{"Alt-Svc":["h3=\":443\"; ma=2592000"],"Server":["Caddy"]}}
{"level":"error","ts":1693576485.3023813,"logger":"http.log.access.log0","msg":"handled request","request":{"remote_ip":"redacted","remote_port":"27931","client_ip":"redacted","proto":"HTTP/2.0","method":"GET","host":"vault.fqdn.com","uri":"/api/config","headers":{"User-Agent":["Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.5615.50 Safari/537.36"],"Sec-Fetch-Mode":["cors"],"Device-Type":["7"],"Bitwarden-Client-Version":["2023.5.1"],"Accept-Language":["en-US"],"Sec-Ch-Ua-Platform":["\"macOS\""],"Sec-Fetch-Site":["cross-site"],"Sec-Fetch-Dest":["empty"],"Sec-Ch-Ua":["\"Not:A-Brand\";v=\"99\", \"Chromium\";v=\"112\""],"Bitwarden-Client-Name":["desktop"],"Accept":["application/json"],"Cache-Control":["no-store"],"Accept-Encoding":["gzip, deflate, br"],"Pragma":["no-cache"],"Sec-Ch-Ua-Mobile":["?0"]},"tls":{"resumed":false,"version":772,"cipher_suite":4865,"proto":"h2","server_name":"vault.fqdn.com"}},"bytes_read":0,"user_id":"","duration":0.015668939,"size":0,"status":502,"resp_headers":{"Server":["Caddy"],"Alt-Svc":["h3=\":443\"; ma=2592000"]}}`

**Comment by BlackDex** [comment]
502 means bad-gateway, which could be that Vaultwarden isn't running, or the container isn't accessible from the caddy container.

**Reply by giteke** [reply]
Thanks, that was indeed the case. However, I'm at a loss on why vaultwarden is suddenly crashing. 

Here are the logs:

`vaultwarden    | thread 'main' panicked at 'OS can't spawn worker thread: Operation not permitted (os error 1)', /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.31.0/src/runtime/scheduler/multi_thread/worker.rs:447:13

vaultwarden    | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

vaultwarden exited with code 101
`

I thought it could be a docker engine issue so I upgraded but unfortunately that didn't resolve it.

**Reply by stefan0xC** [reply]
What OS and docker version are you using? Can you add the output of `docker info`?

**Reply by tigerkzr** [reply]
same issue with the panic on server:latest image

docker log:
```
[2023-09-01 11:33:32.165][panic][ERROR] thread 'main' panicked at 'Error running migrations: QueryError(DieselMigrationName { name: "2023-06-17-200424_create_auth_requests_table", version: MigrationVersion("20230617200424") }, DatabaseError(Unknown, "Can't create table `vaultwarden`.`auth_requests` (errno: 150 \"Foreign key constraint is incorrectly formed\")"))': src/db/mod.rs:471
   0: vaultwarden::init_logging::{{closure}}
   1: std::panicking::rust_panic_with_hook
   2: std::panicking::begin_panic_handler::{{closure}}
   3: std::sys_common::backtrace::__rust_end_short_backtrace
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: vaultwarden::db::mysql_migrations::run_migrations
   8: vaultwarden::db::DbPool::from_config
   9: vaultwarden::main::{{closure}}
  10: tokio::runtime::park::CachedParkThread::block_on
  11: tokio::runtime::context::runtime::enter_runtime
  12: tokio::runtime::runtime::Runtime::block_on
  13: vaultwarden::main
  14: std::sys_common::backtrace::__rust_begin_short_backtrace
  15: std::rt::lang_start::{{closure}}
  16: std::rt::lang_start_internal
  17: main
  18: <unknown>
  19: __libc_start_main
  20: _start
```

docker info:
```
Client: Docker Engine - Community
 Version:    24.0.5
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.20.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 91
  Running: 90
  Paused: 0
  Stopped: 1
 Images: 88
 Server Version: 24.0.5
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 nvidia runc
 Default Runtime: nvidia
 Init Binary: docker-init
 containerd version: 8165feabfdfe38c65b599c4993d227328c231fca
 runc version: v1.1.8-0-g82f18fe
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.4.12-060412-generic
 Operating System: Ubuntu 23.04
 OSType: linux
 Architecture: x86_64
 CPUs: 48
 Total Memory: 251.8GiB
 Name:
 ID: APOX:PAL4:UHQR:HOK3:ZNHC:FBO6:4B2U:BBZ4:EWFJ:2EKS:QBEE:7JCJ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Reply by stefan0xC** [reply]
Can you check with `sqlite3` if your `data/db.sqlite3` file is okay? (I.e. by running [`PRAGMA integrity_check;`](https://www.sqlite.org/pragma.html#pragma_integrity_check)?)

**Comment by BlackDex** [comment, marked as answer]
See #3827, that has the exact same issue.

**Reply by tigerkzr** [reply]
that appears to be the interaction with older docker version..  I'm on 24.0.5 and am getting the panic that prevents the container from starting.  See my reply above for error and docker info

**Reply by BlackDex** [reply]
In that case see #3824

**Reply by giteke** [reply]
Thanks, I tried to updating to to the latest docker following the official docks but ran into issues with the compose. and buildx plugins not installing. However, I'm on Ubuntu 16.04 so I'll bring it current and see if that sorts things out.

Edit: Issue resolved by updating Ubuntu and installing the latest docker version
